frontend/packages/dev-console/src/components/topology/topology-utils.ts

```diff
@@ -0,0 +1,436 @@
+    const resourceData = this.filterBasedOnActiveApplication(this.resources[targetDeployment].data);
+
+    _.forEach(resourceData, (deploymentConfig) => {
+      deploymentConfig.kind = targetDeploymentsKind;
+      const dcUID = _.get(deploymentConfig, 'metadata.uid');
+
+      const replicationController = this.getReplicationController(
+        deploymentConfig,
+        targetDeployment,
+      );
+      const dcPods = this.getPods(replicationController);
+      const service = this.getService(deploymentConfig);
+      const route = this.getRoute(service);
+      const buildConfigs = this.getBuildConfigs(deploymentConfig);
+      // list of resources in the
+      const nodeResources = [deploymentConfig, replicationController, service, route, buildConfigs];
+      // populate the graph Data
+      this.createGraphData(deploymentConfig);
+      // add the lookup object
+      const deploymentsLabels = _.get(deploymentConfig, 'metadata.labels') || {};
+      const deploymentsAnnotations = _.get(deploymentConfig, 'metadata.annotations') || {};
+      this.topologyData.topology[dcUID] = {
+        id: dcUID,
+        name:
+          deploymentsLabels['app.kubernetes.io/instance'] ||
+          _.get(deploymentConfig, 'metadata.name'),
+
+        type: 'workload',
+        resources: _.map(nodeResources, (resource) => {
+          resource.name = _.get(resource, 'metadata.name');
+          return resource;
+        }),
+        data: {
+          url: !_.isEmpty(route.spec) ? getRouteWebURL(route) : null,
+          editUrl: deploymentsAnnotations['app.openshift.io/edit-url'],
+          builderImage: deploymentsLabels['app.kubernetes.io/name'],
+          isKnativeResource: this.isKnativeServing(deploymentConfig, 'metadata.labels'),
+          donutStatus: {
+            pods: _.map(dcPods, (pod) =>
+              _.merge(_.pick(pod, 'metadata', 'status', 'spec.containers'), {
+                id: _.get(pod, 'metadata.uid'),
+                name: _.get(pod, 'metadata.name'),
+                kind: 'Pod',
+              }),
+            ),
+          },
+        },
+      };
+    });
+    return this;
+  }
+
+  private getSelectorsByService() {
+    const allServices = _.keyBy(this.resources.services.data, 'metadata.name');
+    const selectorsByService = _.mapValues(allServices, (service) => {
+      return new LabelSelector(service.spec.selector);
+    });
+    return selectorsByService;
+  }
+
+  /**
+   * filter data based on the active application
+   * @param data
+   */
+  private filterBasedOnActiveApplication(data) {
+    const PART_OF = 'app.kubernetes.io/part-of';
+    if (!this.application) {
+      return data;
+    }
+    return data.filter((dc) => {
+      return dc.metadata.labels[PART_OF] && dc.metadata.labels[PART_OF] === this.application;
+    });
+  }
+
+  /**
+   * get the route information from the service
+   * @param service
+   */
+  private getRoute(service: ResourceProps): ResourceProps {
```

Comment:
It looks like a lot of the logic from the project status page is duplicated here. We should find a way to share code.

Reply:
Yes it was probably used as inspiration because at the time we didn't want to refactor the console code to help with a merge down the road. We can tackle the refactoring easier once in the same repo. Although I do not like having multiple components in a single file or having utilities depend on files with ui components. Some cleaner separation of utils and components would be good.